src/ini.cr

```diff
@@ -19,4 +19,21 @@ class INI
     end
     ini
   end
+
+  # Generates an INI-style configuration from a given hash.
+  #
+  # ```
+  # INI.build({"foo" => {"a" => "1"}}, " ") # => "[foo]\na = 1\n\n"
+  # ```
+  def self.build(ini : Hash(String, Hash(String, String)), space : String = "") : String
```

Comment:
Is a configurable `space` really useful? And why is it not a single whitespace by default?

Reply:
I do this because there are INI files with and without spaces around `=`. It's not clear what is the good default. For example, systemd services has without spaces and `php.ini` has one space.

Reply:
`space` should be IMO `" "` (space!) from default.

Reply:
This method (and the overload below) should also take `NamedTuple(String, String)` as an _ini_ argument.

Reply:
Hard to decide between no spaces and one space. By looking around, the specifications of [Microsoft](https://msdn.microsoft.com/en-us/library/ms717987(v=vs.85).aspx) and [Cloanto](https://cloanto.com/specs/ini/) tend to use no spaces, but [Apache](https://commons.apache.org/proper/commons-configuration/apidocs/org/apache/commons/configuration2/INIConfiguration.html) and [PHP](https://secure.php.net/manual/en/configuration.file.php) use one around `=`.
Anyone has others *official* sources for INI specifications?

Reply:
@Sija I'd also expect `space` to be `" "` by default, but I guess it makes more sense to have it empty to avoid compatibility issues. The reasoning is that an INI parser which recognizes whitespace should also be fine if it is left out, but maybe not necessarily the other way around. A naive implementation might not expect whitespace at those places.
This is just a minor tendency, though.

Reply:
@Sija How can we do this? `NamedTuple` can only be instantiated with named arguments, which we don't know in advance. Perhaps we could use a macro for `NamedTuple` somehow?

Reply:
@j8r on the second thought, the whole type restriction for _ini_ argument looks wrong. It should be possible to pass different types of values and not only `String`. Dictionaries should be accepted as both `Hash` as well as `NamedTuple`.

Reply:
@j8r You can loop over a NamedTuple with `#each` exactly like a Hash.
@Sija `.parse` returns only String. But for build, it's probably nice feature so users don't have to stringify every value beforehand. And it doesn't require effort here, so it's a good addition to make the Hash type-agnostic.

Reply:
In fact, removing the type restriction from the method signature, the code should instantly work for named tuples and other payload types than String.

Reply:
We can add `Boolean, Number, String` according to the different implementations listed in [Wikipedia](https://en.wikipedia.org/wiki/INI_file). `Array` isn't always supported.

Reply:
No, just remove all type restrictions. This way, you can pass in any type and the builder just prints the result of `to_s`. This is the most developer-friendly solution.

Reply:
@straight-shoota I know, but adding jt `NamedTuple(String, String) ` raises `can only instantiate NamedTuple with named arguments`.

By the way regardless of the types, the mapping should follow the `Key(Value, Key(Key, Value)` structure.

That's OK, we will just need to add verifications and raise custom errors to avoid incorrect INI files if we accept everything.

Reply:
Just do it like this:
```crystal
module INI
  def self.build(io : IO, ini, space : String = "")
    ini.each do |section, contents|
      io << '[' << section << "]\n"
       contents.each do |key, value|
         io << key << space << '=' << space << value << '\n'
       end
       io.puts
     end
  end
end

INI.build(STDOUT, {"main": {foo: "bar", baz: 12, regex: /regex/}})
```
No need for type restrictions or custom errors. If a data structure that does not provide a matching interface, it will automatically result in a compile time error.